File: LeetCode Solutions/June LeetCoding Challenge/Stone Game VII.py
class Solution:
    def stoneGameVII(self, stones: List[int]) -> int:
        dpCurrent, dpLast = [0]*len(stones), [0]*len(stones)
        for i in range(len(stones)-2, -1, -1):
            total = stones[i]
            dpLast, dpCurrent = dpCurrent, dpLast
            for j in range(i+1, len(stones)):
                total += stones[j]
                dpCurrent[j] = max(total - stones[i] - dpLast[j], total - stones[j] - dpCurrent[j-1])
        return dpCurrent[-1]

File: LeetCode Solutions/LeetCoding Challenge/June 2020 LeetCoding Challenge/Stone Game VII.cpp
class Solution {
public:
    int stoneGameVII(vector<int>& stones) {
        vector<int> dpCurrent(stones.size()), dpLast(stones.size());
        for(int i=stones.size()-2; ~i; i--){
            int total = stones[i];
            dpLast.swap(dpCurrent);
            for(int j=i+1; j<stones.size(); j++){
                total += stones[j];
                dpCurrent[j] = max(total - stones[i] - dpLast[j], total - stones[j] - dpCurrent[j-1]);
            }
        }
        return dpCurrent[stones.size()-1];
    }
};